3rdparty/mshadow/CMakeLists.txt

```diff
@@ -59,6 +59,9 @@ endif()
 if(USE_CUDNN)
   target_compile_definitions(mshadow INTERFACE MSHADOW_USE_CUDNN)
 endif()
+if(USE_CUTENSOR)
+  target_compile_definitions(mshadow INTERFACE MSHADOW_USE_CUTENSOR)
+endif()
```

Comment:
Please add the option at the mxnet/CMakeLists.txt, analogous to USE_CUDNN. I guess it should default to ON? Is cutensor included by default in cuda 11?

Reply:
Thanks, will add. But will keep as OFF by default.
cutensor is not part of cuda 11, so user needs to download independently: [cutensor](https://developer.nvidia.com/cutensor)

Reply:
I'm a bit concerned that we add too many things which depend on optional dependencies that are determined at compile time. This make releasing mxnet more and more difficult. Can we fine a neater way? Like either saying that cutensor is now a required dependency if cudnn and/or cuda are enabled or put them into the external operator space that is currently being developed? We are creating more and more branching within our build system that is impossible to test properly. 

Another topic is of course the whole licensing topic.

Reply:
there shouldn't be a licensing concern given that it's an optional dependency

Reply:
No that doesn't really make a difference, we have to consider them either way.

We now again have competing operator implementations which are chosen at compile time - numpyGpu vs cutensor. This creates another situation where from a user perspective, we are adding another magic knob which improves performance but is pretty hidden. As a product, we should make mxnet smart enough to figure these things out at runtime - run with cutensor if available and fall back to numpy implementation (our own) if not available. But adding even more compile time branching sounds totally user-unfriendly in my opinion. 

In the end, this might totally worth for a benchmark scenario or a situation where somebody knows exactly what's happening in every detail of mxnet, but the standard user should get a good Performance without having to know every single detail of every single operator and which build flags to choose - this is a more fundamental thing.

Reply:
While I agree that we should have good defaults for build options, I don't think we should mix any of the build options together. We can't foresee how downstream users want to use certain features, and easy customization is essential in accommodating as much of the customized usage as possible. Many successful projects have a large number of build options for this reason: https://github.com/opencv/opencv/blob/master/CMakeLists.txt

Reply:
I agree that users should be able to choose, but I'm arguing that we should rather do that at run time opposed to having every nitty gritty option at compile time.

Reply:
having the option to select features at runtime means that all features must be built into the binary, which is based on the assumption that such build is feasible.

Reply:
Not all, but we have to start somewhere. We can't make every feature a compile time option and then say that it's because it's generally not possible. We should start small and incrementally get to that state.

Just this feature as example. We would need an entire different build plus a few full test suite runs to test this particular flavour. And then when talk about publishing, we either have to decide if we enable or disable it - the other option will then get totally forgotten because not many people will bother to compile with every single flag changed. 

If we do it at runtime, we could have systems á la autotuning which determine the best flags (or allows the user to easily toggle), but if we don't start somewhere, we will never get to that point.

Reply:
I know that the c++ world tends towards compile time options because it's just so easy with all the preprocessor statements, but smart loading (e.g. only when the appropriate dependencies and environment are present) and runtime feature toggles are something that can even be done in the c++ world.

Reply:
having intelligent runtime loading would certainly be nice. This doesn't mean that we should force Moises to work on this, and I think whoever works on it should provide an RFC to help provide clarity on how it would work eventually.

Reply:
Sure